pkg/cloud/services/ec2/instances.go

```diff
@@ -262,26 +262,55 @@ func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
 
 	switch {
 	case scope.AWSMachine.Spec.Subnet != nil && scope.AWSMachine.Spec.Subnet.ID != nil:
+		if failureDomain != nil {
+			criteria := filter.EC2.SubnetID(aws.StringValue(scope.AWSMachine.Spec.Subnet.ID))
+			subnets, err := s.getFilteredSubnets(criteria)
+			if err != nil {
+				return "", errors.Wrapf(err, "failed to filter subnets for criteria %q", criteria)
+			}
+
+			if len(subnets) == 0 {
+				record.Warnf(scope.AWSMachine, "FailedCreate",
+					"Failed to create instance: subnet with id %q not found", aws.StringValue(scope.AWSMachine.Spec.Subnet.ID))
+				return "", awserrors.NewFailedDependency(
+					fmt.Sprintf("failed to run machine %q, subnet with id %q not found",
+						scope.Name(),
+						aws.StringValue(scope.AWSMachine.Spec.Subnet.ID),
```

Comment:
Can we introspect the subnet information from the AWSCluster object instead? I'm wary of adding an additional EC2 API call if we can avoid it.

Reply:
Do you mean something like this `s.scope.Subnets().FindByID("id")` ?

Reply:
Yup. It might be we need to persist more information in AWSCluster status to enable this.

Reply:
I see the subnets being populated [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/a229052b19a7b4d976f312a6bf6a42a9ce09b5d8/pkg/cloud/services/network/subnets.go#L148), but I am not entirely sure, what extra data need to go in there. Is there some documentation that I can refer to?
Should we also modify `getFilteredSubnets` to look for subnets in status, and fallback to ec2 api if it's not there?

Reply:
I think there should be everything needed already persisted:

https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/a229052b19a7b4d976f312a6bf6a42a9ce09b5d8/pkg/cloud/services/network/subnets.go#L264

Does that have everything needed @shysank ?

Reply:
I've made the changes, PTAL.